player/video.c

```diff
@@ -665,20 +665,16 @@ double calc_average_frame_duration(struct MPContext *mpctx)
 // effective video FPS. If this is not possible, try to do it for multiples,
 // which still leads to an improved end result.
 // Both parameters are durations in seconds.
-static double calc_best_speed(double vsync, double frame, int max_factor)
+static double calc_best_speed(double vsync, double frame,
+                              double max_change, int max_factor)
 {
     double ratio = frame / vsync;
-    double best_scale = -1;
-    double best_dev = INFINITY;
     for (int factor = 1; factor <= max_factor; factor++) {
         double scale = ratio * factor / rint(ratio * factor);
-        double dev = fabs(scale - 1);
-        if (dev < best_dev) {
-            best_scale = scale;
-            best_dev = dev;
-        }
+        if (fabs(scale - 1) <= max_change)
+            return scale;
     }
-    return best_scale;
+    return -1;
```

Comment:
Just minor nit here, but I think you can just return 0 instead in this case. That way you wouldn't have to change the logic in `find_best_speed` to handle the `-1` value.

Reply:
That would influence the result because `num` still gets incremented.

In the case of variable frame rate, if there is a frame for which no acceptable correction can be found and a bunch of frames for which an acceptable correction can be found, should that one frame still influence the resulting factor? I'm leaning towards no.

Reply:
Oh I see, previously it always incremented, but with the way you're doing it, you might not always want it increment. I think this is fine the way it is. Seems to make sense at least.